course/include/isometry.h

```diff
@@ -0,0 +1,273 @@
+  double &operator[](const int index);
+  const double &operator[](const int index) const;
+
+  double dot(const Vector3 &other) const;
+  Vector3 cross(const Vector3 &other) const;
+  double norm() const;
+  double &x() { return x_; }
+  const double &x() const { return x_; }
+  double &y() { return y_; }
+  const double &y() const { return y_; }
+  double &z() { return z_; }
+  const double &z() const { return z_; }
+
+  static const Vector3 kUnitX;
+  static const Vector3 kUnitY;
+  static const Vector3 kUnitZ;
+  static const Vector3 kZero;
+
+private:
+  double x_, y_, z_;
+};
+
+inline Vector3 operator*(const double &lhs, const Vector3 &rhs) {
+  return (rhs * lhs);
+}
+
+inline std::ostream &operator<<(std::ostream &ss,
+                                const cppcourse::Vector3 &vec) {
+  return ss << "(x: " << vec.x() << ", y: " << vec.y() << ", z: " << vec.z()
+            << ")";
+}
+
+class Matrix3 {
+public:
+  Matrix3() {}
+  Matrix3(const std::initializer_list<double> &rhs);
+  Matrix3(const Matrix3 &rhs);
+  Matrix3(const Vector3 &first, const Vector3 &second, const Vector3 &third) {
+    rows_[0] = first;
+    rows_[1] = second;
+    rows_[2] = third;
+  }
+  Matrix3(Matrix3 &&rhs) {
+    rows_[0] = std::move(rhs.rows_[0]);
+    rows_[1] = std::move(rhs.rows_[1]);
+    rows_[2] = std::move(rhs.rows_[2]);
+  }
+  Matrix3 operator-(const Matrix3 &other) const;
+  Matrix3 operator+(const Matrix3 &other) const;
+  Matrix3 operator*(const double &other) const;
+  Matrix3 operator*(const Matrix3 &other) const;
+  Matrix3 operator*(const Vector3 &other) const;
+  Matrix3 operator/(const Matrix3 &other) const;
+  Matrix3 &operator-=(const Matrix3 &rhs);
+  Matrix3 &operator+=(const Matrix3 &rhs);
+  Matrix3 operator=(const Matrix3 &rhs) { return Matrix3(rhs); };
+  Matrix3 operator=(const Matrix3 &&rhs) { return Matrix3(rhs); };
+  Vector3 &operator[](const int row_n) { return row(row_n); };
+  Vector3 operator[](const int row_n) const { return row(row_n); };
+
+  const Vector3 &row(int index) const { return rows_[index]; }
+  Vector3 &row(int index) { return rows_[index]; }
+  Vector3 col(int index) const {
+    return Vector3(rows_[0][index], rows_[1][index], rows_[2][index]);
+  }
+  bool operator==(const Matrix3 &other) const;
+  static const Matrix3 kIdentity;
+  static const Matrix3 kZero;
+  static const Matrix3 kOnes;
+  double det() const;
+
+private:
+  Vector3 rows_[3];
+};
+
+inline Matrix3 operator*(const double &lhs, const Matrix3 &rhs) {
+  return (rhs * lhs);
+}
+
+inline std::ostream &operator<<(std::ostream &ss, const Matrix3 &vec) {
+  return ss << "[[" << vec.row(0)[0] << ", " << vec.row(0)[1] << ", "
+            << vec.row(0)[2] << "],"
+            << " [" << vec.row(1)[0] << ", " << vec.row(1)[1] << ", "
+            << vec.row(1)[2] << "],"
+            << " [" << vec.row(2)[0] << ", " << vec.row(2)[1] << ", "
+            << vec.row(2)[2] << "]]";
+}
+
+class Vector4 {
```

Comment:
Please split code for the Isometry object into a separate PR. Also, you don't need to define `Vector4` nor `Matrix4`. You should be able to perform all the operations on the isometry matrix using `Vector3` and `Matrix3`. Internally, the 4x4 homogeneus matrix is actually a 3x3 rotation/scale matrix and a 3x1 translation vector with with and additional row.
![IMG_20200312_103956272](https://user-images.githubusercontent.com/17802342/76527548-2fc6b600-644e-11ea-9286-94053152b951.jpg)

Reply:
I'll try to implement that in a different way